### PR TITLE
WIP: add compute-post.

### PR DESCRIPTION
**Pull request by csukuangfj**
Usage:
```bash
$ snowfall net compute-post -m /ceph-fj/model-jit.pt -f exp/data/cuts_test-clean.json.gz -o exp
```

I find that there is one issue with the Torch Scripted module: We have to know the signature of the `forward` function
of the model, also have to know its subsampling factor.

----

Working on `compute-ali` and will submit them together.

### CODE REVIEW

**Comment by csukuangfj** [comment]
I just created a pull-request in Lhotse https://github.com/lhotse-speech/lhotse/pull/319 to add
posteriors to the class `Cut`. The motivation is to reuse the serialization and dataset code from it.

---

Also, I find the alignment information contained in the supervision is too simple, see
https://github.com/lhotse-speech/lhotse/blob/ef7a037426f1b602a54f4d9ea43e711007e85719/lhotse/supervision.py#L24
```python
    symbol: str    
    start: Seconds    
    duration: Seconds
```

Can we move the alignment class from snowfall to lhotse?
https://github.com/k2-fsa/snowfall/blob/bce73304f40c321a6dad809058b12e559962c321/snowfall/tools/ali.py#L20-L28

**Comment by csukuangfj** [comment]
The usage of `compute-ali`:
```bash
$ snowfall  ali compute-ali -l data/lang_nosp -p ./exp/cuts_post.json  --max-duration=500 -o exp
```

File: snowfall/tools/ali.py [hunk 1]
```diff
@@ -196,3 +203,107 @@ def visualize(input: str,
...
+    lang_dir = Path(lang_dir)
+
+    symbol_table = k2.SymbolTable.from_file(lang_dir / 'words.txt')
+    phone_symbol_table = k2.SymbolTable.from_file(lang_dir / 'phones.txt')
+    phone_ids = get_phone_symbols(phone_symbol_table)
+
+    phone_ids_with_blank = [0] + phone_ids
+    ctc_topo = k2.arc_sort(build_ctc_topo(phone_ids_with_blank))
+
+    if not (lang_dir / 'HLG.pt').exists():
```
Comment (pzelasko) [inline snowfall/tools/ali.py:224]: I think this could be refactored to a function and re-used across this script and decode scripts (and possibly others)

```python
def load_or_compile_HLG(lang_dir: Path) -> k2.Fsa: ...
```

File: snowfall/tools/ali.py [hunk 2]
```diff
@@ -196,3 +203,107 @@ def visualize(input: str,
...
+                          labels_disambig_id_start=first_phone_disambig_id,
+                          aux_labels_disambig_id_start=first_word_disambig_id)
+        torch.save(HLG.as_dict(), lang_dir / 'HLG.pt')
+    else:
+        print('Loading pre-compiled HLG')
+        d = torch.load(lang_dir / 'HLG.pt')
+        HLG = k2.Fsa.from_dict(d)
+
+    HLG = HLG.to(device)
+    HLG.aux_labels = k2.ragged.remove_values_eq(HLG.aux_labels, 0)
```
Comment (pzelasko) [inline snowfall/tools/ali.py:246]: what is this line doing? It looks like it's "sparsifying" the aux_labels (word ids) but how does HLG known which labels correspond to which aux_labels after that?

File: snowfall/tools/ali.py [hunk 3]
```diff
@@ -196,3 +203,107 @@ def visualize(input: str,
...
+        indices = torch.argsort(supervision_segments[:, 2], descending=True)
+        supervision_segments = supervision_segments[indices]
+        texts = supervisions['text']  # TODO: remove it
+
+        sf = supervisions['cut'][0].posts.subsampling_factor
+        dense_fsa_vec = k2.DenseFsaVec(nnet_output,
+                                       supervision_segments,
+                                       allow_truncate=sf - 1)
+
+        lattices = k2.intersect_dense_pruned(HLG, dense_fsa_vec, 20.0,
```
Comment (pzelasko) [inline snowfall/tools/ali.py:284]: the pruning related arguments here could be function parameters

File: snowfall/tools/net.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    dataloader = torch.utils.data.DataLoader(dataset,
+                                             batch_size=None,
+                                             sampler=sampler,
+                                             num_workers=1)
+
+    output_dir = Path(output_dir)
+    output_dir.mkdir(exist_ok=True)
+    storage_path = output_dir / 'posts'
+
+    posts_writer = lhotse.NumpyFilesWriter(storage_path=storage_path)
```
Comment (pzelasko) [inline snowfall/tools/net.py:48]: This is going to create a lot of files, not sure if NumpyHdf5Writer is preferable

File: snowfall/tools/ali.py (hunk 1 above)
Comment (csukuangfj) [reply snowfall/tools/ali.py:224]: Thanks! Will refactor it and add options to enable/disable LM rescoring.

**Comment by pzelasko** [comment]
BTW I wonder if we should support piping these programs together, Kaldi-style. Click easily allows doing that with [file type arguments](https://click.palletsprojects.com/en/8.0.x/arguments/#file-arguments).

We could do that by writing/reading JSONL-serialized manifests in a streaming manner. Since most operations on `CutSet` refer to individual operations on `Cut`, this seems feasible without the need to re-write too much code. [There is a function in Lhotse](https://github.com/lhotse-speech/lhotse/blob/ef7a037426f1b602a54f4d9ea43e711007e85719/lhotse/serialization.py#L541) that tries to figure out the right manifest type from a dict, which can be used to parse individual lines (BTW @csukuangfj I just realized that you might need to extend that function to handle the posterior manifests in your Lhotse PR).

WDYT?

**Comment by pzelasko** [comment]
... there is also some code for line-by-line [incremental JSONL writing in Lhotse](https://github.com/lhotse-speech/lhotse/blob/master/lhotse/serialization.py#L189) that could be extended to support this.

**Comment by danpovey** [comment]
This cool; I'm afraid I'm not following it in detail.
Just a reminder; this is more an "experimental direction" at this point.  We'll have to learn from experience whether these kinds of command line utilites are actually a useful thing.

**Comment by pzelasko** [comment]
Fair enough. The idea is to allow sth like: 
```
snowfall net compute-post <some-inputs-args..> - | snowfall net compute-ali - <some-more-args..>
```
but I just realized that with the current way things are done in Lhotse, we would have store the actual arrays/tensors on disk and just pass the manifests around, which might not be optimal. Maybe it's not relevant for now and we can see how to do that in the future, if needed at all.

**Comment by danpovey** [comment]
BTW, I tend to think being able to do something at all tends to be more important than that thing being efficient-- premature optimization being the root of all evil etc., although I did plenty of it in Kaldi.  I don't know  what the optimal solution is here, I am afraid I have not been following this PR closely enough.

**Comment by pzelasko** [comment]
Agreed. But for the record, the full quote is actually: 

> "We should forget about small efficiencies, say about 97% of the time: premature optimization is the root of all evil. Yet we should not pass up our opportunities in that critical 3%."
